cmd/kubelet/app/options/globalflags.go

```diff
@@ -93,7 +92,25 @@ func addCredentialProviderFlags(fs *pflag.FlagSet) {
 
 // addKlogFlags adds flags from k8s.io/klog
 func addKlogFlags(fs *pflag.FlagSet) {
-	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
-	klog.InitFlags(local)
-	fs.AddGoFlagSet(local)
+	// lookup flags in global flag set and re-register the values with our flagset
+	global := flag.CommandLine
+	local := pflag.NewFlagSet(os.Args[0], pflag.ExitOnError)
+
+	// DO NOT call klog.InitFlags. It has the side-effect of resetting
```

Comment:
this seems like something klog should protect against. having to doc `DO NOT call klog.InitFlags` (the method klog provides to register flags) seems super error-prone

I expected a fix in InitFlags to only stomp in the initial values once, like this:

```diff
diff --git a/klog.go b/klog.go
index 887ea62dff7..94d9d915265 100644
--- a/klog.go
+++ b/klog.go
@@ -404,21 +404,32 @@ func init() {
        go logging.flushDaemon()
 }
 
+var initOnce sync.Once
+
 // InitFlags is for explicitly initializing the flags
 func InitFlags(flagset *flag.FlagSet) {
+       initOnce.Do(func() {
+               logging.logDir = ""
+               logging.logFile = ""
+               logging.logFileMaxSizeMB = 1800
+               logging.toStderr = true
+               logging.alsoToStderr = false
+               logging.skipHeaders = false
+               logging.skipLogHeaders = false
+       })
        if flagset == nil {
                flagset = flag.CommandLine
        }
-       flagset.StringVar(&logging.logDir, "log_dir", "", "If non-empty, write log files in this directory")
-       flagset.StringVar(&logging.logFile, "log_file", "", "If non-empty, use this log file")
-       flagset.Uint64Var(&logging.logFileMaxSizeMB, "log_file_max_size", 1800,
+       flagset.StringVar(&logging.logDir, "log_dir", logging.logDir, "If non-empty, write log files in this directory")
+       flagset.StringVar(&logging.logFile, "log_file", logging.logFile, "If non-empty, use this log file")
+       flagset.Uint64Var(&logging.logFileMaxSizeMB, "log_file_max_size", logging.logFileMaxSizeMB,
                "Defines the maximum size a log file can grow to. Unit is megabytes. "+
                        "If the value is 0, the maximum file size is unlimited.")
-       flagset.BoolVar(&logging.toStderr, "logtostderr", true, "log to standard error instead of files")
-       flagset.BoolVar(&logging.alsoToStderr, "alsologtostderr", false, "log to standard error as well as files")
+       flagset.BoolVar(&logging.toStderr, "logtostderr", logging.toStderr, "log to standard error instead of files")
+       flagset.BoolVar(&logging.alsoToStderr, "alsologtostderr", logging.alsoToStderr, "log to standard error as well as files")
        flagset.Var(&logging.verbosity, "v", "number for the log level verbosity")
-       flagset.BoolVar(&logging.skipHeaders, "skip_headers", false, "If true, avoid header prefixes in the log messages")
-       flagset.BoolVar(&logging.skipLogHeaders, "skip_log_headers", false, "If true, avoid headers when openning log files")
+       flagset.BoolVar(&logging.skipHeaders, "skip_headers", logging.skipHeaders, "If true, avoid header prefixes in the log messages")
+       flagset.BoolVar(&logging.skipLogHeaders, "skip_log_headers", logging.skipLogHeaders, "If true, avoid headers when openning log files")
        flagset.Var(&logging.stderrThreshold, "stderrthreshold", "logs at or above this threshold go to stderr")
        flagset.Var(&logging.vmodule, "vmodule", "comma-separated list of pattern=N settings for file-filtered logging")
        flagset.Var(&logging.traceLocation, "log_backtrace_at", "when logging hits line file:N, emit a stack trace")
```

Reply:
I agree that we should also fix klog.
This was meant as a quick fix since we have to also fix klog and then pull in the updated dep.
@vincepri was going to work on the fix to klog.